File: extras/Python/tests/loopback_transport.py
"""In-memory LoopbackTransport for unit testing.

Port of C# ``LoopbackTransport.cs``. Uses a ``collections.deque`` as a
byte buffer. :meth:`simulate_receive` injects data as if it arrived from
the embedded side (fires :attr:`data_received`).
"""
from __future__ import annotations

import threading
from collections import deque

from cmd_messenger.transport.transport import Transport


class LoopbackTransport(Transport):
    """In-memory transport stub for host-side tests — no serial port needed."""

    def __init__(self) -> None:
        super().__init__()
        self._buffer: deque[bytes] = deque()
        self._lock = threading.Lock()
        self._connected = False

    def connect(self) -> bool:
        self._connected = True
        return True

    def disconnect(self) -> bool:
        self._connected = False
        return True

    def is_connected(self) -> bool:
        return self._connected

    def read(self) -> bytes:
        with self._lock:
            if self._buffer:
                return self._buffer.popleft()
        return b""

    def write(self, data: bytes) -> None:
        if not data:
            return
        with self._lock:
            self._buffer.append(bytes(data))
        self.data_received(self)

    def simulate_receive(self, data: bytes) -> None:
        """Inject bytes as if they arrived from the remote side."""
        if not data:
            return
        with self._lock:
            self._buffer.append(bytes(data))
        self.data_received(self)

    def dispose(self) -> None:
        self._connected = False

File: extras/CSharp/Tests/CommandMessenger.Tests/LoopbackTransport.cs
using System;
using System.Collections.Concurrent;
using CommandMessenger.Transport;

namespace CommandMessenger.Tests
{
    /// <summary>
    /// In-memory ITransport implementation for unit testing.
    /// Write() enqueues bytes; Read() dequeues them.
    /// Raises DataReceived after every Write().
    /// </summary>
    public class LoopbackTransport : ITransport
    {
        private readonly ConcurrentQueue<byte[]> _buffer = new();
        private bool _connected;

        public event EventHandler DataReceived;

        public bool Connect()
        {
            _connected = true;
            return true;
        }

        public bool Disconnect()
        {
            _connected = false;
            return true;
        }

        public bool IsConnected() => _connected;

        public byte[] Read()
        {
            if (_buffer.TryDequeue(out var data))
                return data;
            return Array.Empty<byte>();
        }

        public void Write(byte[] buffer)
        {
            if (buffer == null || buffer.Length == 0) return;
            _buffer.Enqueue((byte[])buffer.Clone());
            DataReceived?.Invoke(this, EventArgs.Empty);
        }

        /// <summary>
        /// Inject bytes as if they arrived from the remote side.
        /// This simulates incoming data without going through Write().
        /// </summary>
        public void SimulateReceive(byte[] data)
        {
            if (data == null || data.Length == 0) return;
            _buffer.Enqueue((byte[])data.Clone());
            DataReceived?.Invoke(this, EventArgs.Empty);
        }

        public void Dispose() { _connected = false; }
    }
}
